LibraBFT/Impl/Handle.agda

```diff
@@ -32,157 +35,148 @@ open import LibraBFT.Yasm.Base
+  runHandler : RoundManager → LBFT Unit → RoundManager × List (LYT.Action NetworkMsg)
+  runHandler st handler = ×-map₂ (outputsToActions {st}) (proj₂ (LBFT-run handler st))
 
 ------------------------------------------------------------------------------
--- TODO : ASK CHRIS : regarding EitherD-run
-
-zzz : EitherD ErrLog ℕ
-zzz = do
-  r ← RightD 0
-  RightD r
-
-zzz' : Either ErrLog ℕ
-zzz' = Right 0
 
-zzz≡zzz' : zzz ≡ RightD 0 → zzz' ≡ Right 0
-zzz≡zzz' ()
+-- This connects the implementation handler to the system model so it can be initialized.
+
+module InitHandler where
+  {-
+  IMPL-DIFF: In Haskell, nodes are started with a filepath of a file containing
+  - number of faults allowed
+  - genesis LedgerInfoWithSignatures
+  - network addresses/name and secret and public keys of all nodes in the genesis epoch
+
+  Main reads/checks that file then calls a network specific 'run' functions (e.g., ZMQ.hs)
+  that setup the network handlers then eventually call 'Start.startViaConsensusProvider'.
+
+  That functions calls 'ConsensusProvider.startConsensus' which returns
+  '(EpochManager, [Output]'.
+
+  'Start.startViaConsensusProvider' then goes on to create and wire up internal communication
+  channels and starts threads.  The most relevant thread starts up
+  `(EpochManager.obmStartLoop epochManager output ...)' to handle the initialization output
+  and then to handle new messages from the network.
+
+  In Agda functions below, assuming no initialization errors,
+  - the 'GenKeyFile.create' function
+    - creates everything that would have been written to disk and then later read
+      to create ValidatorSigner(s), ValidatorVerifier, LIWS, etc.
+  - there is no network transport in Agda
+  - 'initialize' calls 'State.startViaConsensusProvider'
+    when then calls 'ConsensusProvider.startConsensus' which returns
+    '(EpochManager, List Output)', just like Haskell.
+  - Since there is no network, internal channels and threads, this is the end of this process.
+
+  Note: although both the Haskell and Agda version support non-uniform voting, the
+  above initialization process assumes one-vote per peer.
+  -}
+
+  postulate
+    now : Instant
+    pg  : ProposalGenerator
+
+  initialize' : Instant → GenKeyFile.NfLiwsVsVvPe → Either ErrLog (EpochManager × List Output)
+  initialize' now nfLiwsVsVvPe =
+    Start.startViaConsensusProvider
+      now nfLiwsVsVvPe
+      (TxTypeDependentStuffForNetwork∙new pg (StateComputer∙new BlockInfo.gENESIS_VERSION))
+
+  abstract
+    initialize  : Instant → GenKeyFile.NfLiwsVsVvPe → Either ErrLog (EpochManager × List Output)
+    initialize  = initialize'
+    initialize≡ : initialize ≡ initialize'
+    initialize≡ = refl
+
+  mkNfLiwsVsVvPe : BootstrapInfo → ValidatorSigner → GenKeyFile.NfLiwsVsVvPe
+  mkNfLiwsVsVvPe bsi vs = (bsi ^∙ bsiNumFaults , bsi ^∙ bsiLIWS , vs , bsi ^∙ bsiVV , bsi ^∙ bsiPE)
+
+  initEMWithOutput' : BootstrapInfo → ValidatorSigner → Either  ErrLog (EpochManager × List Output)
```

Comment:
As there is no branching here, it probably won't be necessary/beneficial to have an `EitherD` variant, but it doesn't do much harm to keep it.

Reply:
I'll leave it for now.